src/init.cpp

```diff
@@ -1519,93 +1522,125 @@ bool AppInitMain(NodeContext& node)
-                if (!::ChainstateActive().ReplayBlocks(chainparams)) {
-                    strLoadError = _("Unable to replay blocks. You will need to rebuild the database using -reindex-chainstate.").translated;
-                    break;
-                }
-
-                // The on-disk coinsdb is now in a good state, create the cache
-                ::ChainstateActive().InitCoinsCache();
-                assert(::ChainstateActive().CanFlushToDisk());
+                bool failed_chainstate_init = false;
+
+                for (CChainState* chainstate : g_chainman.GetAll()) {
+                    LogPrintf("Initializing chainstate %s\n", chainstate->ToString());
+                    chainstate->InitCoinsDB(
+                        /* cache_size_bytes */ nCoinDBCache,
+                        /* in_memory */ false,
+                        /* should_wipe */ fReset || fReindexChainState);
+
+                    chainstate->CoinsErrorCatcher().AddReadErrCallback([]() {
+                        uiInterface.ThreadSafeMessageBox(
+                            _("Error reading from database, shutting down.").translated,
+                            "", CClientUIInterface::MSG_ERROR);
+                    });
+
+                    // If necessary, upgrade from older database format.
+                    // This is a no-op if we cleared the coinsviewdb with -reindex or -reindex-chainstate
+                    if (!chainstate->CoinsDB().Upgrade()) {
+                        strLoadError = _("Error upgrading chainstate database").translated;
+                        failed_chainstate_init = true;
+                        break;
+                    }
 
-                is_coinsview_empty = fReset || fReindexChainState ||
-                    ::ChainstateActive().CoinsTip().GetBestBlock().IsNull();
-                if (!is_coinsview_empty) {
-                    // LoadChainTip initializes the chain based on CoinsTip()'s best block
-                    if (!::ChainstateActive().LoadChainTip(chainparams)) {
-                        strLoadError = _("Error initializing block database").translated;
+                    // ReplayBlocks is a no-op if we cleared the coinsviewdb with -reindex or -reindex-chainstate
+                    if (!chainstate->ReplayBlocks(chainparams)) {
+                        strLoadError = _("Unable to replay blocks. You will need to rebuild the database using -reindex-chainstate.").translated;
+                        failed_chainstate_init = true;
                         break;
                     }
-                    assert(::ChainActive().Tip() != nullptr);
+
+                    // The on-disk coinsdb is now in a good state, create the cache
+                    chainstate->InitCoinsCache();
+                    assert(chainstate->CanFlushToDisk());
+
+                    is_coinsview_empty = fReset || fReindexChainState ||
```

Comment:
in commit e89fbd3cf0820ee263a94c49751a97a784acb049:

I have a hard time following the logic-change.

* You replace all calls to ChainActive with a for loop over all chainstates
* You only initialize the "ibd" chainstate, which will retain all logic as it was before
* However, is_coinsview_empty will be set to the result of whatever chainstate was processed last, and then used (in the next loop) on whatever chainstate is processed first. This seems fragile or at least requires a comment to explain why this can be done.

Reply:
Maybe it's a little hard to see in the diff, but the current value of `is_coinsview_empty` is only ever used within a single loop. In the next loop, the [value is reset](https://github.com/jamesob/bitcoin/blob/1c173812bac34e263ae49e0bc80e45b437e95e8b/src/init.cpp#L1619-L1620) based on the chainstate being worked on.

It's safe to reset the value on line 1619 because it's not possible for the coinstip's best block to be set to null between this `is_coinsview_empty = ...` assignment and the one in the next loop.

To be extra sure there's no change, I could create a `chainstate_to_coinsview_empty` map and only set the values once, but I'll only do this if you think it's worthwhile.

Reply:
Oh, I missed that the value is reset. Maybe it would help to limit the scope of the variable to make that more clear and avoid this issue accidentally coming up in the future?

Reply:
Fixed by removing the variable entirely and replacing it with a function.